### ISSUE TITLE
Some Matrix4x4 tests are failing on OSX in corefx

### ISSUE THREAD

**Issue by mellinoe** [dotnet/coreclr#2266]
We've noticed that a couple of random tests for our System.Numerics.Matrix4x4 structure have started failing in the past week, only on OSX Release builds. I haven't investigated the issues too deeply yet, but I strongly suspect there is some bug introduced recently in the runtime, because:
- The library and tests haven't changed in a long time
- The tests pass on all other combinations of frameworks, OS's, etc.
- They were passing unchanged a couple of weeks ago :smile: 

Note that Matrix4x4 is a very large structure (16 floats = 64-bytes wide), and I think there have been some changes to how structures are passed around on other platforms. The issue is only happening on OSX release builds, not Linux or any other configuration. Does anyone recall recent changes that might be causing this issue in that particular configuration? If I had to guess, I would say it's something to do with the fact that it is such a large structure, but again, I haven't investigated deeply yet.

Tests failing:

[Matrix4x4CreateFromYawPitchRollTest2](http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/osx_release_tst_prtest/79/console)

```
07:00:32    System.Numerics.Tests.Matrix4x4Tests.Matrix4x4CreateFromYawPitchRollTest2 [FAIL]
07:00:32       Yaw:-265 Pitch:155 Roll:-405
07:00:32       Expected: True
07:00:32       Actual:   False
07:00:32       Stack Trace:
07:00:33             at System.Numerics.Tests.Matrix4x4Tests.Matrix4x4CreateFromYawPitchRollTest2()
```

[Source code for CreateFromYawPitchRoll()](https://github.com/dotnet/corefx/blob/master/src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs#L1160)

[Matrix4x4DecomposeTest01](http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/osx_release_tst_prtest/19/console)

```
08:54:36    System.Numerics.Tests.Matrix4x4Tests.Matrix4x4DecomposeTest01 [FAIL]
08:54:36       Matrix4x4.Decompose did not return expected value. Expected:<10, 20, 30> actual:<10, 20.19675, 30>.
08:54:36       Expected: True
08:54:36       Actual:   False
08:54:36       Stack Trace:
08:54:36             at System.Numerics.Tests.Matrix4x4Tests.DecomposeTest(Single yaw, Single pitch, Single roll, Vector3 expectedTranslation, Vector3 expectedScales)
08:54:36             at System.Numerics.Tests.Matrix4x4Tests.Matrix4x4DecomposeTest01()
```

[Source code for Decompose()](https://github.com/dotnet/corefx/blob/master/src/System.Numerics.Vectors/src/System/Numerics/Matrix4x4.cs#L1493)

**Comment by tzwlai** [comment]
@RussKeldorph, does this look like it could be related to the JIT?

**Comment by RussKeldorph** [comment]
@tzwlai probably makes sense for us to look.  @LLITCHEV @CarolEidt FYI.

**Comment by sivarv** [comment]
@RussKeldorph - This is an issue on OSX, should we move it to 'future'?

**Comment by LLITCHEV** [comment]
@sivarv I am not 100% it is issue on the Mac only. (In terms of the Jit there should be no differences on Linux, Mac OS or any other *nix OS.) I this investigating at least will not hurt.

**Comment by LLITCHEV** [comment]
@mellinoe I could not repro this. There were fixed to the jit that could potentially fixed this. It seems it is passing now on the MacOS CI (at least one of the test is not disabled - the first one.) Please retest and update the bug.

**Comment by RussKeldorph** [comment]
@mellinoe Please re-open if this is still repro'ing.

**Comment by mellinoe** [comment]
We're still seeing intermittent failures in these tests in our OSX Release test runs in corefx. Note that these tests are deterministic (no random elements), but are only failing intermittently.

**Comment by janvorli** [comment]
I wonder if the failures can be related to the physical hardware the test run ends up running on, in other words if the problem is similar to the issue #2687.
@LLITCHEV this is what I have mentioned today.

**Comment by sivarv** [comment]
It is unlikely that issue #2687 could be the cause of this issue because only very recently (1/15) Carol has enabled SSE2 SIMD intrinsics on Linux.  AVX support for SIMD intrinsics is not turned on for *nix yet.  Also, my understanding is that these tests were failing even before that.

**Comment by mellinoe** [comment]
Yes, the tests have been failing for a little while, longer than the SIMD feature was turned on. Also, these failures are occurring on OSX, which I don't believe has SIMD enabled yet? I think only Linux has been enabled recently.

Another thing to note is that the test failures are in `Matrix4x4`, which doesn't have any special code-gen as far as I know. It uses `Vector3/Vector4` internally in some places which might have special recognition, but I actually don't think the code paths used by the test hit that.

**Comment by CarolEidt** [comment]
It is my understanding that OSX uses the same codegen as Linux. So, I believe that SIMD is enabled there as well. But, as you say, this was failing before the SIMD feature was enabled.

**Comment by LLITCHEV** [comment]
@CarolEidt Yes. There is no any difference in terms of codegen (or in any other way) between Linux and MacOS by the Jit. I wonder if the failures are on the same machine? It is really perplexing that a deterministic test will fail intermittently. And I'd expect for it to fail on Linux too.

**Comment by mellinoe** [comment]
@LLITCHEV was able to run the tests on the same machine that we saw one of the CI failures, and we still weren't able to get a repro, unfortunately. We were also using the same bits that the CI machine used when it failed, taken from the shared Jenkins location. Since this seems to be such an obscure problem, I'm moving to Future but leaving it open for now. Hopefully we can get some better information in the future regarding why this happens.

Main questions that we haven't solved:

* Why does it only happen on OSX if the code-gen should be identical between Linux/OSX/(others)?
* Why does it affect the Matrix4x4 tests? (big structure?)
* Why is it intermittent if the tests are deterministic?

**Comment by stephentoub** [comment]
Failed again here: http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/osx_release_tst/277/console

**Comment by mellinoe** [comment]
@LLITCHEV I believe these are all on the same machine, including Stephen's new one today. If I'm reading the Jenkins log correctly, all of the failures are on [dci-mac-bld-6](http://dotnet-ci.cloudapp.net/computer/dci-mac-bld-6). Is that the one that we had tested on? @mmitche do you remember which machine we had pulled to do the test?

**Comment by LLITCHEV** [comment]
@mellinoe The machines that we pulled to try to repro the were 131.107.19.71 and 74.30.245.235. I believe the second was the one that you last repro occurred. After many runs it didn't happen even once. @mmitche Could you please confirm that dci-mac-0bld-6 is the same machine?

**Comment by mmitche** [comment]
06 is http://131.107.19.118/

**Comment by mellinoe** [comment]
There were a couple more repros that @stephentoub noted in the corefx bug:

http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/osx_release_tst_prtest/144/console
http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/osx_release_tst_prtest/145/console

Those failures happened on [dci-mac-bld-4](http://dotnet-ci.cloudapp.net/computer/dci-mac-bld-4) and [dci-mac-bld-6](http://dotnet-ci.cloudapp.net/computer/dci-mac-bld-1), so I think we can rule out this being specific to a particular machine. It has happened on 1, 4, and 6 at the very least.

**Comment by mellinoe** [comment]
@LLITCHEV Perhaps we can try to repro this by running the full set of tests, same as is done in CI? It may be that this issue only arises under the pressure of running all the tests in succession.

**Comment by RussKeldorph** [comment]
@LLITCHEV look into this again if you can.  Intermittent failures on OSX sounds more like some other issues I've heard about than codegen.  I wonder if perhaps this has stopped repro'ing by now.

**Comment by stephentoub** [comment]
>  I wonder if perhaps this has stopped repro'ing by now.

We disabled the tests in https://github.com/dotnet/corefx/issues/5834 because they were taking out too many runs randomly.

**Comment by RussKeldorph** [comment]
@stephentoub Right.  My wondering was was referring to the hypothetical situation of the test actually being run.  Is it worth re-enabling again to confirm?  If the repro rate is low enough, that may be the only practical option.  I believe we should certainly get to the bottom of this before RTM if the test is still failing.

**Comment by LLITCHEV** [comment]
@RussKeldorph @stephentoub I think re-enabling the test makes sense. In worse case we will know where we stand.

**Comment by RussKeldorph** [comment]
I hate not knowing what was going on here, but if it isn't repro'ing anymore, my best guess is that it was  a manifestation of one of the non-JIT OSX-specific issues that was found and fixed separately.  If anyone has a better course of action than "close this and re-open if and when it repros again," feel free to re-open and explain.

**Comment by RussKeldorph** [comment]
Another repro: dotnet/corefx#9409.

**Comment by RussKeldorph** [comment]
@stephentoub @mellinoe Has this been quiet because there have been no additional repros or because you guys have given up hope we'll make traction on it?

**Comment by stephentoub** [comment]
> Has this been quiet because there have been no additional repros

There have been additional repros.  It looks like we have multiple related issues.  e.g. I just logged one here yesterday:
https://github.com/dotnet/corefx/issues/4833#issuecomment-247228222

**Comment by RussKeldorph** [comment]
This turns out to be pretty easy to repro on my MacBook Pro.  Just running the following in a tight bash loop will repro it in <~50 iterations.

```
./corerun xunit.console.netcore.exe System.Numerics.Vectors.Tests.dll -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapp1.0tests -notrait category=nonosxtests
```

Thus far, I've only been able to repro it when I *don't* pass the `-parallel none` switch to xUnit, though I'll leave a loop running overnight to see.  Right now, though, since the codegen is the same between Linux and OS X, this really smells more like a PAL or OS issue related to threading and/or FP context save/restore.  Will continue to dig as best I can for now.

@janvorli @kouvel @gkhanna79

**Comment by RussKeldorph** [comment]
I verified that all the relevant generated code for the test is identical between OS X and Ubuntu.  This also repros regardless of whether AVX is used (which makes sense since it first occurred before we enabled SIMD on Unix).  Because this seems to repro only when multithreading is involved (no repro after 24 hours under xunit -parallel none) and only on OS X, I'm inclined to believe this is either a PAL problem or something wrong with OS X itself.

@gkhanna79 FYI

**Comment by gkhanna79** [comment]
@JohnChen0 will take a look at this.

CC @janvorli

**Comment by mellinoe** [comment]
@gkhanna79 @JohnChen0 Did we ever find any new information about this? We still see repro's of this in some of our test runs.

**Comment by JohnChen0** [comment]
I haven't been able to make much progress on this issue. I have been able to repro it locally, but since the repro is non-deterministic, it is hard to investigate what caused it.

**Comment by janvorli** [comment]
@RussKeldorph when you have reproduced this error, were the differences between the expected and actual numbers also small like in the reported Matrix4x4DecomposeTest01 or were there cases when they were completely off?

**Comment by JohnChen0** [comment]
I had one repro of Matrix4x4DecomposeTest01, with the following message: `Matrix4x4.Decompose did not return expected value Expected:<2, 3, 4> actual:<2, 3, NaN>.` I would call that completely off.

**Comment by danmosemsft** [comment]
@JohnChen0 any progress?

**Comment by RussKeldorph** [comment]
@janvorli Sorry I missed your question.  I confirm what @JohnChen0 saw.  NaNs were common, though I can't remember if they occurred every time.  I can attempt to recreate the repro.

**Comment by janvorli** [comment]
@RussKeldorph no problem. I have looked through the activation injection code that I suspected and it seems that we have the same problem there we had on Linux in the past. I don't have Mac available now, but could you or @JohnChen0 try to make a simple one line change in PAL to see if it fixes the problem?
The change would be here:
https://github.com/dotnet/coreclr/blob/master/src/pal/src/exception/machexception.cpp#L1583
The `CONTEXT_FULL | CONTEXT_SEGMENTS` would be changed to `CONTEXT_CONTROL | CONTEXT_INTEGER | CONTEXT_SEGMENTS`
The point is that we capture context including the floating point registers when we "suspend" a thread for GC. When we are returning from the suspension, we restore the xmm registers, but we are not saving restoring the full ymm ones on machines that have them.
But there is no need to capture and restore the floating point stuff in the GC suspension, since GC doesn't use xmm registers.

**Comment by CarolEidt** [comment]
I don't think it is correct that GC doesn't use xmm registers. It calls some math library methods that use the xmm registers. I know this because when we had some issues with firmware problems with testing OS support for AVX, the GC was running into the issue due to its use of AVX via these crt methods. @Maoni0 can probably confirm this.

**Comment by JohnChen0** [comment]
I'll try out the change @janvorli suggested, and see if it fixes the issue.

**Comment by JohnChen0** [comment]
The tests still fail intermittently after changing https://github.com/dotnet/coreclr/blob/master/src/pal/src/exception/machexception.cpp#L1583 as suggested.

**Comment by Maoni0** [comment]
@CarolEidt, could you please elaborate on these issues GC hit? I had not heard about them.

**Comment by CarolEidt** [comment]
Unfortunately, I no longer have the original email thread on this topic, but it came up about a year and a half ago. There was some bad firmware that was wrongly reporting the OS bits of the mask, and both clr and the crt were directly querying those bits and occasionally getting the wrong answer. Around the same time the crt introduced a new version of some of the math functions (pow, if I recall correctly) that were dynamically deciding whether to use AVX, and suddenly caused the clr to start failing, even with COMPlus_EnableAVX=0 (which turns of AVX code in the JIT). I believe (again, if I recall correctly) that the uses of pow() were in the GC. I can't find such uses now, so either it has changed, or it was somewhere else in the vm.

**Comment by Maoni0** [comment]
my guess would be it was somewhere in the vm.

**Comment by janvorli** [comment]
@JohnChen0 could you please try one more thing? It would be helpful to know if the issue is related to activation injection, that means GC. 
So, my idea is to try to print xmm0..7 from pContext right after we extract the CONTEXT here:
https://github.com/dotnet/coreclr/blob/master/src/pal/src/exception/machexception.cpp#L1586
And then again right before we restore the context here
https://github.com/dotnet/coreclr/blob/master/src/pal/src/exception/machexception.cpp#L1515

I would be print each xmm register as two 64 bit hex numbers and also print with the current thread id. On OSX, the only thread id you can get is this:
```
    uint64_t tid;
    pthread_threadid_np(pthread_self(), &tid);
```
Then when we hit the problem, we should be able to see if the values got screwed somehow.

Or we can save the xmm registers or the whole copy of the context to a TLS variable in the first place and then compare the xmm registers at the second one and abort if they differ.

**Comment by JohnChen0** [comment]
@janvorli  I've instrumented the code to dump Xmm register contents. Here is a dump before a test failure:

    Saving Context, tid=446944
      Xmm0 = 0000000000000000 4040000040000000
      Xmm1 = 0000000000000000 0000000040800000
      Xmm2 = 0000000000000000 3fefffffbf708fb1
      Xmm3 = 0000000000000000 00000000324cde2e
      Xmm4 = 0000000000000000 00000000264d8ab9
      Xmm5 = 0000000000000000 00000000b43bbd2e
      Xmm6 = 0000000000000000 000000003f708fb1
      Xmm7 = 0000000000000000 00000000beaf1d4b
    Restoring Context, tid=446951
      Xmm0 = 0000000000000000 4040000040000000
      Xmm1 = 0000000000000000 0000000040800000
      Xmm2 = 0000000000000000 3fefffffbf708fb1
      Xmm3 = 0000000000000000 00000000324cde2e
      Xmm4 = 0000000000000000 00000000264d8ab9
      Xmm5 = 0000000000000000 00000000b43bbd2e
      Xmm6 = 0000000000000000 000000003f708fb1
      Xmm7 = 0000000000000000 00000000beaf1d4b
       System.Numerics.Tests.Matrix4x4Tests.Matrix4x4DecomposeTest01 [FAIL]
          Matrix4x4.Decompose did not return expected value Expected:<2, 3, 4> actual:<0, 0, 0>.
          Expected: True
          Actual:   False
          Stack Trace:
             /Users/jochen/corefx/src/System.Numerics.Vectors/tests/Matrix4x4Tests.cs(252,0): at System.Numerics.Tests.Matrix4x4Tests.DecomposeTest(Single yaw, Single pitch, Single roll, Vector3 expectedTranslation, Vector3 expectedScales)
             /Users/jochen/corefx/src/System.Numerics.Vectors/tests/Matrix4x4Tests.cs(270,0): at System.Numerics.Tests.Matrix4x4Tests.Matrix4x4DecomposeTest01()
    Finished:    System.Numerics.Vectors.Tests
    
The register contents are identical. I did notice the thread ID changed. Not sure if that's expected.

**Comment by janvorli** [comment]
The thread id change is correct. The context is read on a different thread (but from the context of the suspended target thread) and restored on the target thread.
@JohnChen0 Does the activation injection invocation that manifests itself with these prints look related to the issue? What I mean, when the issue happens, it is always there and when it doesn't happen, it is not there or at least not right before the failure?

**Comment by CarolEidt** [comment]
> @janvorli I've instrumented the code to dump Xmm register contents. Here is a dump before a test failure:

Earlier in the thread, someone Can you dump the full ymm register contents? That would be where I would expect the context restore issue to lie.

**Comment by JohnChen0** [comment]
I'm not sure where the ymm registers are supposed to be saved inside `CONTEXT`. There aren't any fields with name similar to that. Is it by some other name, or is it actually missing in `CONTEXT`?

**Comment by CarolEidt** [comment]
The ymm registers need to be part of `CONTEXT` if any operation that is done during a suspend can modify them.

**Comment by janvorli** [comment]
Actually, it seems I've just found it - x86_avx_state64_t

**Comment by janvorli** [comment]
I've created #8066 to fix this issue, assigned to @JohnChen0

**Comment by gkhanna79** [comment]
We will use this bug to track fixing for 1.2.0 and use #8066 to track fixing it for 1.0.X and 1.1.X.

**Comment by janvorli** [comment]
@gkhanna79 actually, I've created #8066 for 1.2.0 and #8067 for the porting to 1.0.x

**Comment by JohnChen0** [comment]
I've figured out the cause of this issue. It's actually not related to YMM registers, but is due to XMM8 through XMM15 registers. The vector code generated by JIT uses all 16 XMM registers, but currently we only save 8 XMM registers on Mac. What's worse, we restore all 16 XMM registers, so XMM8 through XMM15 are restored from random memory, and is guaranteed to be corrupted. Thus if the capture/restore occurs during a segment of code where XMM8 through XMM15 are in use, we get bad result.

While debugging this, I also found another minor bug in the context save code. The `context.cpp` file in PAL has the following kind of code at multiple places:

    if (lpContext->ContextFlags & CONTEXT_ALL_FLOATING)

This doesn't actually work as expected. On x64, `CONTEXT_ALL_FLOATING` is defined as `(CONTEXT_AMD64 | 0x8L)`. When the caller doesn't want to save floating point registers, the `0x8L` bit isn't set, but the `CONTEXT_AMD64` bit is always set, thus the `if` expression evaluates to `true` even when floating point registers are not requested. This is a minor issue. We can save registers needlessly, but there isn't any correctness problem.

I'll submit a fix to save XMM8 through XMM15 soon, and file another issue to fix the `if` statements.

**Comment by mellinoe** [comment]
Thanks for tracking this one down! This has been randomly hitting corefx tests for a while and it's always been a complete mystery for me.

One question: Is it ever necessary for us to preserve the YMM registers as well? Obviously we aren't seeing any problems or test failures related to it right now, but I would guess a similar scenario could potentially occur there. Basically I'm just asking for clarification about a point @CarolEidt made above:

> The ymm registers need to be part of CONTEXT if any operation that is done during a suspend can modify them.

**Comment by janvorli** [comment]
@mellinoe we should preserve the YMM registers on OSX too. I guess we are just lucky that the C++ runtime doesn't use them.
@JohnChen0 would you mind adding that as described in the #8066?

**Comment by JohnChen0** [comment]
Sounds good. I'll add the code to preserve the YMM registers.

**Comment by JohnChen0** [comment]
Fixed with PR #8078 (current release), #8099 (release 1.1.0), and #8098 (release 1.0.*)
